treecorr/catalog.py

```diff
@@ -1513,59 +1534,63 @@ def set_patch(data, patch_col):
-            all_hdus = set(all_hdus + [hdu])
+            x_ext = infile.choose_extension(self.config, 'x_ext', num, ext)
+            y_ext = infile.choose_extension(self.config, 'y_ext', num, ext)
+            z_ext = infile.choose_extension(self.config, 'z_ext', num, ext)
+            ra_ext = infile.choose_extension(self.config, 'ra_ext', num, ext)
+            dec_ext = infile.choose_extension(self.config, 'dec_ext', num, ext)
+            r_ext = infile.choose_extension(self.config, 'r_ext', num, ext)
+            patch_ext = infile.choose_extension(self.config, 'patch_ext', num, ext)
+            w_ext = infile.choose_extension(self.config, 'w_ext', num, ext)
+            wpos_ext = infile.choose_extension(self.config, 'wpos_ext', num, ext)
+            flag_ext = infile.choose_extension(self.config, 'flag_ext', num, ext)
+            g1_ext = infile.choose_extension(self.config, 'g1_ext', num, ext)
+            g2_ext = infile.choose_extension(self.config, 'g2_ext', num, ext)
+            k_ext = infile.choose_extension(self.config, 'k_ext', num, ext)
+            all_exts = [x_ext, y_ext, z_ext,
+                        ra_ext, dec_ext, r_ext,
+                        patch_ext,
+                        w_ext, wpos_ext, flag_ext,
+                        g1_ext, g2_ext, k_ext]
+            col_by_ext = dict(zip(all_cols,all_exts))
+            all_exts = set(all_exts + [ext])
             all_cols = [c for c in all_cols if c != '0']
 
             data = {}
             # Also, if we are only reading in one patch, we should adjust s before doing this.
             if self._single_patch is not None:
                 if patch_col != '0':
-                    data[patch_col] = fits[patch_hdu][patch_col][s]
+                    data[patch_col] = infile.read(patch_ext, patch_col, s)
                     all_cols.remove(patch_col)
                     set_patch(data, patch_col)
+                    end1 = infile.row_count(patch_ext, patch_col)
                 elif self._centers is not None:
                     pos_cols = [x_col, y_col, z_col, ra_col, dec_col, r_col]
                     pos_cols = [c for c in pos_cols if c != '0']
                     for c in pos_cols:
                         all_cols.remove(c)
-                    for h in all_hdus:
-                        use_cols1 = [c for c in pos_cols if col_by_hdu[c] == h]
-                        data1 = fits[h][use_cols1][s]
+                    for h in all_exts:
+                        use_cols1 = [c for c in pos_cols if col_by_ext[c] == h]
+                        data1 = infile.read(h, use_cols1, s)
                         for c in use_cols1:
                             data[c] = data1[c]
+                    end1 = infile.row_count(h, c)
                     set_pos(data, x_col, y_col, z_col, ra_col, dec_col, r_col)
                 use = self._get_patch_index(self._single_patch)
                 self.select(use)
                 if isinstance(s,np.ndarray):
```

Comment:
Similarly put a `#pragma: no branch` if you don't want to bother with the mock test.